src/main/java/net/syscon/elite/repository/jpa/model/CourtEvent.java

```diff
@@ -115,19 +126,48 @@ public void setEventDateTime(final LocalDateTime dateTime) {
         this.startTime = dateTime;
     }
 
-    @PrePersist
-    private void carryOverPreExistingActiveChargesOnCreation() {
-        if (offenderCourtCase != null) {
-            log.debug("Adding pre-existing active charges to new court event for court case '{}", offenderCourtCase.getId());
+    private void add(final CourtEventCharge charge) {
+        charges.add(charge);
+    }
 
-            offenderCourtCase.getCharges().stream().filter(OffenderCharge::isActive).forEach(this::addPreExisting);
+    static class OnCreate {
+        /**
+         * On court event creation on NOMIS existing active charges are applied to the new court event if associated with a court case.
+         */
+        @PrePersist
+        void applyActiveCourtCaseChargesFor(final CourtEvent event) {
+            event.getOffenderCourtCase().ifPresent(courtCase -> {
+                        courtCase.getCharges(OffenderCharge::isActive).forEach(charge ->
+                                event.add(CourtEventCharge.builder()
+                                        .offenderCharge(charge)
+                                        .courtEvent(event)
+                                        .build())
+                        );
+
+                        if (!event.getCharges().isEmpty()) {
+                            log.debug("Carried over '{}' active charge(s) for court case '{}' to court event '{}'", event.getCharges().size(), courtCase.getId(), event.getId());
+                        }
+                    }
+            );
         }
     }
 
-    private void addPreExisting(final OffenderCharge charge) {
-        charges.add(CourtEventCharge.builder()
-                .offenderCharge(charge)
-                .courtEvent(this)
-                .build());
+    @Component
+    static class OnDelete {
+
+        private Clock clock;
+
+        @Autowired
+        void setClock(final Clock clock) {
+            this.clock = clock;
+        }
+
+        @PreRemove
+        void checkIsScheduledFutureEventPriorToRemovalOf(final CourtEvent event) {
+            Objects.requireNonNull(clock, "Clock not set.");
+
+            checkState(event.getEventDateTime().isAfter(LocalDateTime.now(clock)), "Court hearing '%s' cannot be deleted as its start date/time is in the past.", event.getId());
+            checkState(event.getEventStatus().isScheduled(), "Court hearing '%s' must be in a scheduled state to delete.", event.getId());
```

Comment:
What exception do you end up getting when the state isn't satisfied?

Reply:
You do get this exception but is wrapped in another exception on instance type DataAccessException.  We should be able to deal with this in higher layers.